---
title: "R Notebook"
output: html_notebook
---


```{r}
#####
#> TO CHANGE:
#> - modify which version of the DGE objects you want to load, i.e. with cut-off (Treat_xxx) or not (glmQL_xxx). Currently loading both and assigning glmQL_ as the ones in use
#>
#####

### ENTER ROOT DIR PATH BELOW, or keep the renv/project rootdir
rootdir <- getwd()
Robjectsdir <- paste0(rootdir, "/Robjects")
Graphsdir <- paste0(rootdir, "/graphs")
# make heatmaps graph folder
if (!dir.exists(file.path(Graphsdir, "Heatmaps"))) {
  dir.create(file.path(Graphsdir, "Heatmaps"))
}
Heatdir <- paste0(Graphsdir, "/Heatmaps")
# make volcano plots graph folder
if (!dir.exists(file.path(Graphsdir, "Volcanoes"))) {
  dir.create(file.path(Graphsdir, "Volcanoes"))
}
Volcanodir <- paste0(Graphsdir, "/Volcanoes")

### ENTER PATH WITH KALLISTO FOLDERS BELOW
datapath <- paste0(rootdir, "/input")
# sample names for all samples
sample_list <- dir(file.path(datapath))

#setting root directory
knitr::opts_knit$set(root.dir = rootdir)
# options(max.print = .Machine$integer.max, scipen = 999, stringsAsFactors = F, dplyr.summarise.inform = F) # avoid truncated output in R console and scientific notation
# Set seed to ensure reproducibility in any function sthat require randomness
set.seed(123456)
suppressMessages({
  library(edgeR)    # for DGE analysis
  library(pheatmap) # for heatmap graphs
  library(org.Dm.eg.db)    # Dmel db for GO terms/Kegg pathways
  library(ggplot2)  # graphing module
  library(ggrepel)  # for labelling stuff - volcano plots
  # library(tidyverse) # includes ggplot2, for data visualisation. dplyr, for data manipulation.
  library(RColorBrewer) # for a colourful plot
  library(DGEsurf)  # bespoke functions to probe in code
})
```

```{r loading relevant values}
# if (!exists("y")){
#   y <- readRDS(file = paste0(Robjectsdir, "/y.rds"))
# }
# if (!exists("fit")){
#   fit <- readRDS(file = paste0(Robjectsdir, "/fit.rds"))
# }
if (!exists("s2c")){
  s2c <- readRDS(file = paste0(Robjectsdir, "/s2c.rds"))
  design <- s2c$design
}
if (!exists("t2g")){
  t2g <- readRDS(file = paste0(Robjectsdir, "/t2g_dmel.rds"))
}
if (!exists("logCPM")){
  logCPM <- readRDS(file = paste0(Robjectsdir, "/logCPM.rds"))
  avlogCPM <- readRDS(file = paste0(Robjectsdir, "/avlogCPM.rds"))
}
if (!exists("glmQL_res")){
  glmQL_res <- readRDS(file = paste0(Robjectsdir, "/glmQL_res_txi.rds"))
  glmQL_decided <- readRDS(file = paste0(Robjectsdir, "/glmQL_decided_txi.rds"))
  Treat_res <- readRDS(file = paste0(Robjectsdir, "/Treat_res_txi.rds"))
  Treat_decided <- readRDS(file = paste0(Robjectsdir, "/Treat_decided_txi.rds"))
  QL_res <- glmQL_res
  QL_decided <- glmQL_decided
}
if (!exists("contrasts_list")){
  contrasts_list <- readRDS(file = paste0(Robjectsdir, "/contrasts.rds"))
}


```

### Heatmaps of top 30 DE transcripts in each contrast, mapped onto all contrasts (or cotrasts of same timepoint)

```{r heatmaps top 30 genes for each contrast looped}
# retrieving factors from s2c for annotation in heatmap
groupings <- s2c[,2:3]
rownames(groupings)<-s2c$sample

#looping over every contrast to get top DE transcripts, then make heatmap ocer every sample
for (i in 1:length(QL_res)){
  name <- names(QL_res)[[i]]
  #ordering results of DE testing function by their pval, smallest to largest
  o <- order(QL_res[[name]]$table$PValue)
  loggedcpm = NULL
  #only keeping the logCPM values for the top x DE transcripts
  loggedcpm <- logCPM[o[1:30],]
  print("e")
  # creating heatmap with logCPM values from each sample of the 30 most DE transcripts
  pheatmap(loggedcpm, annotation_col = groupings,
           filename = paste0(Heatdir,"/Heatmap top 30 - ",name,".pdf"),
           main =paste0("Top 30 DE genes - ",name),fontsize_row = 7
           )
}

```

```{r heatmap single contrast}
### Modify the pheatmap inputs if you want to make heatmap over specific group of samples

name <- wcontrast()
o <- order(QL_res[[name]]$table$PValue)
loggedcpm = NULL
#only keeping the logCPM values for the top x DE transcripts
loggedcpm <- logCPM[o[1:30],]
# creating heatmap with logCPM values from each sample of the 30 most DE transcripts
pheatmap(loggedcpm[,colnames(loggedcpm)%in% sample_list], annotation_col = grouping,
         filename = paste0(Heatdir,"/Heatmap top 30 - ",name,".pdf"),
         main =name,fontsize_row = 7
         )
```

# Volcano plots 

```{r volcano plot - making and saving plots, eval=FALSE, include=FALSE}
# https://biostatsquid.com/volcano-plots-r-tutorial/

# global settings for ggplot themes
theme_set(theme_classic(base_size = 20) +
            theme(
              axis.title.y = element_text(face = "bold", margin = margin(0,20,0,0), size = rel(1.5), color = 'black'),
              axis.title.x = element_text(hjust = 0.5, face = "bold", margin = margin(20,0,0,0), size = rel(1.5), color = 'black'),
              plot.title = element_text(hjust = 0.5),
              legend.text = element_text(size = rel(0.5))
            ))

for (i in 1:length(QL_decided)){
  name <- names(QL_decided)[i]
  
  # topTags to get BH statistics from differential expression test.
  # graphing the no-cutoff DE result since I'm using volcano plot to visualize significance
  df_nolfc <- QL_decided[[i]]
  # adding differentially expressed column to results
  df_nolfc$table$diffexpressed <- "NO"
  # if log2Foldchange > 0.26 and pvalue < 0.05, set as "UP" (lfc(1.2)=0.26)
  df_nolfc$table$diffexpressed[df_nolfc$table$logFC > 0.26 & df_nolfc$table$FDR < 0.05] <- "UP"
  # if log2Foldchange < -0.26 and pvalue < 0.05, set as "DOWN"
  df_nolfc$table$diffexpressed[df_nolfc$table$logFC < -0.26 & df_nolfc$table$FDR < 0.05] <- "DOWN"
  # adding column in table to plot gene symbol for top 30 DE transcripts
  df_nolfc$table$delabel <- c(df_nolfc$table$ext_gene[1:30],
                              rep(NA,length(df_nolfc$table$ext_gene)-30))
  # calling plot
  ggplot(data = df_nolfc, aes(x = logFC, 
                            y = -log10(FDR), 
                            col = diffexpressed,
                            label = delabel)) +
    # adding dashed lines to show significance limits
    geom_vline(xintercept = c(-0.26, 0.26), col = "gray", linetype = 'dashed') +
    geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') + 
    geom_point(size =1) + 
    scale_color_manual(values = c("purple2", "grey", "seagreen3"),
                       labels = c("Downregulated", "Not significant", "Upregulated")) + 
    coord_cartesian( xlim = c(-10, 10)) + # since some genes can have minuslog10padj of inf, we set these limits
    labs(color = "Expression",
         x = expression("log"[2]*"FC"), y = expression("-log"[10]*"FDR"),
         title = name) + 
    scale_x_continuous(breaks = seq(-10, 10, 2)) + # to customise the breaks in the x axis
    theme(legend.title = element_text(size = rel(1.5)),
          title = element_text(size = rel(0.5))) +
    geom_text_repel(max.overlaps = Inf) # plotting labels
  ggsave(filename = paste0(Volcanodir, "/volcano_",name, ".pdf"))
}

```


```{r volcano plot single plot}

name <- wcontrast()
# topTags to get BH statistics from differential expression test.
# graphing the no-cutoff DE result since I'm using volcano plot to visualize significance
df_nolfc <- QL_decided[[name]]
# adding differentially expressed column to results
df_nolfc$table$diffexpressed <- "NO"
# if log2Foldchange > 0.26 and pvalue < 0.05, set as "UP" (lfc(1.2)=0.26)
df_nolfc$table$diffexpressed[df_nolfc$table$logFC > 0.26 & df_nolfc$table$FDR < 0.05] <- "UP"
# if log2Foldchange < -0.26 and pvalue < 0.05, set as "DOWN"
df_nolfc$table$diffexpressed[df_nolfc$table$logFC < -0.26 & df_nolfc$table$FDR < 0.05] <- "DOWN"
# adding column in table to plot gene symbol for top 30 DE transcripts
df_nolfc$table$delabel <- c(df_nolfc$table$ext_gene[1:30],
                            rep(NA,length(df_nolfc$table$ext_gene)-30))
# calling plot
ggplot(data = df_nolfc, aes(x = logFC, 
                          y = -log10(FDR), 
                          col = diffexpressed,
                          label = delabel)) +
  # adding dashed lines to show significance limits
  geom_vline(xintercept = c(-0.26, 0.26), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') + 
  geom_point(size =1) + 
  scale_color_manual(values = c("purple2", "grey", "seagreen3"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) + 
  coord_cartesian( xlim = c(-10, 10)) + # since some genes can have minuslog10padj of inf, we set these limits
  labs(color = "Expression",
       x = expression("log"[2]*"FC"), y = expression("-log"[10]*"FDR"),
       title = name) + 
  scale_x_continuous(breaks = seq(-10, 10, 2)) + # to customise the breaks in the x axis
  theme(legend.title = element_text(size = rel(1.5)),
        title = element_text(size = rel(0.5))) +
  geom_text_repel(max.overlaps = Inf) # plotting labels
ggsave(filename = paste0(Volcanodir, "/volcano_",name, ".pdf"))
```

