---
title: "R Notebook"
output: html_notebook
---

```{r}
#####
#> TO CHANGE
#> - modify filenames to which ORA and GSEA are saved, according to if its glmQL or Treat
#> - Currently, ORA and GSEA for GO terms is for all three ontologies. Change 'ont = all' to 'BP', CC, or MF, if separation is desired.
#####
### ENTER ROOT DIR PATH BELOW, or keep the renv/project rootdir
rootdir <- getwd()
Robjectsdir <- paste0(rootdir, "/Robjects")

#setting root directory
knitr::opts_knit$set(root.dir = rootdir)
# options(max.print = .Machine$integer.max, scipen = 999, stringsAsFactors = F, dplyr.summarise.inform = F) # avoid truncated output in R console and scientific notation
# Set seed to ensure reproducibility in any function sthat require randomness
set.seed(123456)
suppressMessages({
  library(clusterProfiler) # ORA/GSEA
  library(org.Dm.eg.db)    # Dmel db for GO terms/Kegg pathways
  library(DOSE)            # for some graphing functions (facet_grid)
  library(enrichplot)      # Graphing functions (pairwise_termsim)
  library(ggplot2)  # graphing module
  library(openxlsx) # to save enrichment analysis results
})
```


```{r loading relevant values}

if (!exists("y")){
  y <- readRDS(file = paste0(Robjectsdir, "/y_gene.rds"))
  if (dim(y$counts)[1]<20000){
    flygenes <- rownames(y$counts)
    t2ged <- t2g[t2g$ens_gene %in% flygenes & !duplicated(t2g$ens_gene),]
    # Creating annotation dataframe to be entered in DGEList y
    # Turning t2ged into data.table object
    t2ged <- as.data.table(t2ged)
    # reordering t2ged by same order as kallisto counts object through transcript col
    t2ged<-t2ged[order(factor(t2ged$ens_gene,levels = flygenes))]
  }else{
    transcripts <- rownames(y$counts)
    t2ged <- t2g[t2g$target_id %in% transcripts & !duplicated(t2g$target_id),]
    #Warning: this removes duplicates, i.e. only gives one row of IDs for each transcript (2248 transcripts duplicates exist where they map to multiple gene IDs).
    #!!! INCREDIBLY IMPORTANT PART => REORDER t2ged IN THE SAME ORDER AS THE COUNTS MATRIX, SO TRANSCRIPTS ARE LINKED TO THE CORRECT GENE IDS
    # Turning t2ged into data.table object
    t2ged <- as.data.table(t2ged)
    # reordering t2ged by same order as kallisto counts object through transcript col
    t2ged<-t2ged[order(factor(t2ged$target_id,levels = rownames(cts.scaled)))]
  }
}
if (!exists("glmQL_res")){
  glmQL_res <- readRDS(file = paste0(Robjectsdir, "/glmQL_res_txi.rds"))
  glmQL_decided <- readRDS(file = paste0(Robjectsdir, "/glmQL_decided_txi.rds"))
  Treat_res <- readRDS(file = paste0(Robjectsdir, "/Treat_res_txi.rds"))
  Treat_decided <- readRDS(file = paste0(Robjectsdir, "/Treat_decided_txi.rds"))
  QL_res <- glmQL_res
  QL_decided <- glmQL_decided
}
if (!exists("contrasts_list")){
  contrasts_list <- readRDS(file = paste0(Robjectsdir, "/contrasts.rds"))
}


```


# Over-representation analysis

```{r ORA - Preparing DE transcript vector, include=FALSE}
# following loop gets vector of DE genes in two lists, either of FBgn IDs or of KEGG IDs (ncbi-geneids)
DET_vecs <- list()
DET_vecs_K <- list()
for (i in 1:length(QL_decided)){
  name <- names(QL_decided)[i]
  # extract pre-specified (n) number of transcripts, here taking every transcript which went through DE analysis. Selection of DE transcripts is done a bit further down.
  DE_results_edgeR <- QL_decided[[i]]
  # first, we transform the results table to a data frame so that we see the results
  # table directly when accessing it through the name "res"
  DE_results_edgeR <- as.data.frame(DE_results_edgeR)
  
  # names(DE_results_edgeR) <- make.names(names(DE_results_edgeR))

  # rename column that contains adjusted p-values: rename padj to p_adj
  colnames(DE_results_edgeR)[colnames(DE_results_edgeR) == "FDR"] <- "p_adj"

  # getting indicator of transcripts with DE FDR result <0.05. Note that when using results from treat_res, the pvalue is relative to them being significantly DE accounting for the cut-off.
  ind_differentially_expressed <- ((!is.na(DE_results_edgeR$p_adj)) & (DE_results_edgeR$p_adj<0.05))
  # using this indicator, we extract the vector of differentially expressed genes 
  # from the results of differential expression analysis
  DE_df <- DE_results_edgeR[ind_differentially_expressed,]
  # Retrieving list of gene IDs for DE transcripts of this contrast
  DET_vecs[[name]] <- DE_df$ens_gene
  # making DE transcripts vector for KEGG analysis, removing genes without 
  # ENTREZ ID (==ncbi-geneid) since they won't be usable for KEGG enrichment analysis
  DET_vecs_K[[name]] <- DE_df$entrez_gene[!is.na(DE_df$entrez_gene)]
}
```

```{r ORA - Running analysis loop, include=FALSE}
ORA_results_GO <- list()
ORA_results_K <- list()
# looping over each contrast's DE transcript vector and running ORA
for (i in 1:length(DET_vecs)){
  name <- names(DET_vecs[i])
  ORA_results_GO[[name]] <- enrichGO(gene = DET_vecs[[name]],
                           OrgDb = org.Dm.eg.db,
                           keyType = "ENSEMBL", # ftbrXXXX
                           ont = "all",
                           # readable = TRUE,   # translates gene ID into gene symbol.
                           minGSSize = 10,       # default is 10. 
                           maxGSSize = 800,
                           # pAdjustMethod = "none"
                           )
  DET_vecs_K[[i]] <- DET_vecs_K[[i]][!duplicated(DET_vecs_K[[i]])]
  ORA_results_K[[name]] <- enrichKEGG(gene = DET_vecs_K[[name]],
                                 organism = "dme",
                                 keyType = "ncbi-geneid", #ENTREZID, numbers.
                                 minGSSize = 10)
  print(name)
}

saveRDS(ORA_results_GO, file = paste0(Robjectsdir, "/ORA_results_GO_glmQL-txi.rds"))
saveRDS(ORA_results_K, file = paste0(Robjectsdir, "/ORA_results_K_glmQL-txi.rds"))
```



```{r ORA - Running analysis single, include=FALSE}
# call "wcontrast()" function to assign the contrast to analyse for
name <- wcontrast()
ORA_results_GO[[name]] <- enrichGO(gene = DET_vecs[[name]],
                         OrgDb = org.Dm.eg.db,
                         keyType = "ENSEMBL", # FBgnxxx
                         ont = "all",
                         # readable = TRUE,   # translates transcript ids to symbol.
                         minGSSize = 10,       # default is 10. 
                         maxGSSize = 800,
                         # pAdjustMethod = "none"
                         )
ORA_results_K[[assigned]] <- enrichKEGG(gene = DET_vecs_K[[assigned]],
                               organism = "dme",
                               keyType = "ncbi-geneid", #ENTREZID, numbers.
                               minGSSize = 10)

# head(ORA_results_GO, 20)
# head(ORA_results_K, 20)
# ORA_GO_CC <-ORA_results_GO[[assigned]][ORA_results_GO[[assigned]]@result[["ONTOLOGY"]]=="CC"]
# ORA_GO_MF<-ORA_results_GO[[assigned]][ORA_results_GO[[assigned]]@result[["ONTOLOGY"]]=="MF"]
```

# Geneset enrichment analysis
```{r GSEA - ranking}
# modify ranking method if desired. 
#1 - ranking by log(FC); 
#2 - ranking by (-)log10(p-value)*sign[log(FC)]
rankvecs <- list()
rankvecs_k <- list()
ranking_method="1"
for (i in 1:length(QL_decided)){
  # extract pre-specified (n) number of transcripts, here taking every transcript which went through DE analysis, since I'm doing GSEA and looking at the whole set of genes in the contrast
  DE_results_edgeR_ENSEMBLTRANS <- QL_decided[[i]]
  name <- names(QL_decided)[i]
  # first, we transform the results table to a data frame so that we see the results
  # table directly when accessing it through the name "res"
  DE_results_edgeR_ENSEMBLTRANS <- as.data.frame(DE_results_edgeR_ENSEMBLTRANS)
  # rename column that contains adjusted p-values: rename padj to p_adj
  DE_results_edgeR_ENSEMBLTRANS <- dplyr::rename(DE_results_edgeR_ENSEMBLTRANS, 
                                           p_adj = FDR)
  # create vector that contains the value of the gene-level ranking metric for each gene. 
  # using LogFC here directly instead of through adjusted pval, simply because for contrasts of interest a majority of transcripts that were tested for DE have pval=1, leading to identical rank for a majority of genes which messes up the GSEA calculations for all genes (including those with significant p-values). Since GSEA method requires further checking of the significance and strength of DE for the genes in significantly enriched terms/pathways, and is really about understanding if there's a trend towards similar expression changes for genes in a term/pathway, this method is acceptable.
  if (ranking_method=="1"){
    rankvec_edgeR_ENSEMBLTRANS <- setNames(DE_results_edgeR_ENSEMBLTRANS$logFC,
                                           rownames(DE_results_edgeR_ENSEMBLTRANS))
  }else if (ranking_method=="2"){
    # OG method below
    rankvec_edgeR_ENSEMBLTRANS <- sign(DE_results_edgeR_ENSEMBLTRANS$logFC)*(-1)*log10(DE_results_edgeR_ENSEMBLTRANS$p_adj)
    # assign respective transcript ID to each value in the vector
    names(rankvec_edgeR_ENSEMBLTRANS) <- rownames(DE_results_edgeR_ENSEMBLTRANS)
  }
  # making vector for KEGG analysis
  rankvec_edgeR_Entrez <- rankvec_edgeR_ENSEMBLTRANS
  # assigning ENTREZID to separate vector for KEGG analysis
  names(rankvec_edgeR_Entrez) <- DE_results_edgeR_ENSEMBLTRANS$entrez_gene
  # sorting the vectors in descending order
  rankvec_edgeR_ENSEMBLTRANS <- sort(rankvec_edgeR_ENSEMBLTRANS, decreasing=TRUE)
  rankvec_edgeR_Entrez <- sort(rankvec_edgeR_Entrez, decreasing=TRUE)
  rankvecs[[name]] <- rankvec_edgeR_ENSEMBLTRANS
  rankvecs_k[[name]] <- rankvec_edgeR_Entrez
}

```


```{r GSEA - running analysis}
gsego_res <- list()
gsekegg_res <- list()
# looping over every ranking vector to run GSEA for all contrasts

for (i in 1:length(rankvecs)){
  name <- names(QL_res)[i]
  gsego_res[[name]] <- gseGO(geneList = rankvecs[[name]],
                     ont = "ALL",
                     OrgDb = org.Dm.eg.db,
                     keyType = "ENSEMBL",
                     minGSSize = 5,
                     eps = 0, 
                     seed = TRUE) # set seed for reproducibility
  # removing duplicate ENTREZ IDS from multiple transcripts mapping to one gene
  # since fgsea behaves differently with duplicate entries.
  # rankvecs_k[[name]] <- rankvecs_k[[name]][!duplicated(names(rankvecs_k[[name]]))]
  gsekegg_res[[name]] <- gseKEGG(geneList = rankvecs_k[[name]], 
                     organism = "dme", 
                     seed = TRUE, 
                     keyType = "ncbi-geneid", 
                     minGSSize = 5,
                     pvalueCutoff = 0.05,
                     # pAdjustMethod = "none"
                     )
}

saveRDS(gsego_res, file = paste0(Robjectsdir, "/GSEA_results_GO_glmQL-txi.rds"))
saveRDS(gsekegg_res, file = paste0(Robjectsdir, "/GSEA_results_K_glmQL-txi.rds"))
```

