---
title: "R Notebook"
output: html_notebook
---

```{r}
### ENTER ROOT DIR PATH BELOW, or keep the renv/project rootdir
rootdir <- getwd()
Robjectsdir <- paste0(rootdir, "/Robjects")
Graphsdir <- paste0(rootdir, "/graphs")
BVdir <- paste0(Graphsdir, "/BioVar")

#setting root directory
knitr::opts_knit$set(root.dir = rootdir)
# options(max.print = .Machine$integer.max, scipen = 999, stringsAsFactors = F, dplyr.summarise.inform = F) # avoid truncated output in R console and scientific notation
# Set seed to ensure reproducibility in any function sthat require randomness
set.seed(123456)
suppressMessages({
  library(edgeR)    # for DGE analysis
  library(ggplot2)  # graphing module
})
```

```{r loading relevant Robjects}
if (!exists("y")){
  y <- readRDS(file = paste0(Robjectsdir, "/y_gene.rds"))
}
if (!exists("fit")){
  fit <- readRDS(file = paste0(Robjectsdir, "/fit.rds"))
}
if (!exists("s2c")){
  s2c <- readRDS(file = paste0(Robjectsdir, "/s2c.rds"))
  design <- s2c$design
}
if (!exists("t2g")){
  t2g <- readRDS(file = paste0(Robjectsdir, "/t2g_dmel.rds"))
}
if (!exists("logCPM")){
  logCPM <- readRDS(file = paste0(Robjectsdir, "/logCPM.rds"))
}
if (!exists("contrasts")){
  contrasts <- readRDS(file = paste0(Robjectsdir, "/contrasts.rds"))
}

```
### Running DGE analysis, producing relevant graphs, and saving data to Robject file
Differential Expression testing

Using glmTreat function to perform Differential expression testing with a log-fold change cutoff of 1.5. With input being a DGEGLM object from glmQLFit(), the testing method used is the same as that used in glmQLFTest, that is a quasi-likelihood F-test, but with the null hypothesis tested as the absolute value of the logFC of each transcript being equal to or below 1.5.

```{r DE testing}
#####
#> To change:
#> - After getting the final number of DE transcripts for each contrast, determine if a significance cut-off is desired and to what level.
#> Remember that any cut-off will bias the resulting significantly DE transcripts for transcripts with potentially low overall expression but relatively high differences in expression in the contrast. Transcripts with higher average expression will need larger increase/decrease in absolute expression values in the contrast to pass the filter, which can hide DE transcripts of genes that are usually highly expressed. But also, getting very large number of DE transcripts (in proportion to total number of transcripts in genome/counts) does inform the need for a cut-off.
#>
#####

# takes as input the DGEGLM object fit previously made when calling glmQLFit. Default FDR is at 5%. Using glmTreat to allow for log-fold-cutoff to focus on more biologically relevant transcripts with higher fold-change in the tested contrast
Treat_res <- list()
for (i in 1:dim(contrasts)[2]){
  Treat_res[[colnames(contrasts)[i]]] <- glmTreat(fit, contrast = contrasts[,i], lfc = log2(1.2))
}
# same method without log-fold cutoff:
glmQL_res <- list()
for (i in 1:dim(contrasts)[2]){
  glmQL_res[[colnames(contrasts)[i]]] <- glmQLFTest(fit, contrast = contrasts[,i])
}
# # additional DE testing options, here to view through ANOVA what is uniquely DE between all treatment groups (including vehicle) for PAO1, pqsR and for all samples.
# Treat_res[["PAO1_aov"]]  <- glmTreat(fit, coef = 2:3, lfc = log2(1.2))
# Treat_res[["psqR_aov"]]  <- glmTreat(fit, coef=4:5, lfc = log2(1.2))
# Treat_res[["all_aov"]]  <- glmTreat(fit, coef=2:5, lfc = log2(1.2))
# glmQL_res[["PAO1_aov"]]  <- glmQLFTest(fit, coef = 2:3)
# glmQL_res[["psqR_aov"]]  <- glmQLFTest(fit, coef=4:5)
# glmQL_res[["all_aov"]]  <- glmQLFTest(fit, coef=2:5)

# changing some names to match with rest of my code
if (!is.null(glmQL_res[[1]]$genes$TranscriptID)){
  for (i in 1:length(glmQL_res)){
    colnames(glmQL_res[[i]]$genes)[colnames(glmQL_res[[i]]$genes) == 'TranscriptID'] <- 'target_id'
    colnames(glmQL_res[[i]]$genes)[colnames(glmQL_res[[i]]$genes) == 'flybaseID'] <- 'ens_gene'
    colnames(glmQL_res[[i]]$genes)[colnames(glmQL_res[[i]]$genes) == 'ENTREZID'] <- 'entrez_gene'
    colnames(glmQL_decided[[i]]$table)[colnames(glmQL_decided[[i]]$table) == 'TranscriptID'] <- 'target_id'
    colnames(glmQL_decided[[i]]$table)[colnames(glmQL_decided[[i]]$table) == 'flybaseID'] <- 'ens_gene'
    colnames(glmQL_decided[[i]]$table)[colnames(glmQL_decided[[i]]$table) == 'ENTREZID'] <- 'entrez_gene'
  }
}

saveRDS(glmQL_res, file = paste0(Robjectsdir, "/glmQL_res_txi.rds"))
saveRDS(Treat_res, file = paste0(Robjectsdir, "/Treat_res_txi.rds"))

```
From above DE testing, number of significantly up/downregulated transcripts, and non-significant transcripts. Looking at 48h inter-host infection contrast here:

```{r Outputting number of DE transcripts, include=FALSE}
glmQL_decided <- list()
glmQL_decided_numbs <- list()

for (i in 1:length(glmQL_res)){
  # gathering number of upreg, NS, and downreg genes per contrast
  glmQL_decided_numbs[[names(glmQL_res)[i]]] <- decideTests(glmQL_res[[i]])
  # multiple testing verification to classify transcripts as significantly upregulated/downregulated or nonsignificant. Tabled information on DGE and its stats in contrast in each item in below list is used for hunt functions
  glmQL_decided[[names(glmQL_res)[i]]] <-  topTags(glmQL_res[[i]], n= nrow(glmQL_res[[i]]))
}

Treat_decided <- list()
Treat_decided_numbs <- list()

for (i in 1:length(Treat_res)){
  Treat_decided_numbs[[names(Treat_res)[i]]] <- decideTests(Treat_res[[i]])
  Treat_decided[[names(Treat_res)[i]]] <- topTags(Treat_res[[i]], n= nrow(Treat_res[[i]]))
}
saveRDS(glmQL_decided, file = paste0(Robjectsdir, "/glmQL_decided_txi.rds"))
saveRDS(Treat_decided, file = paste0(Robjectsdir, "/Treat_decided_txi.rds"))
```

```{r summaries}
# printing out summary of number of upreg, N.S., and downreg genes per contrast, depending on cut-off applied or not.
DGEs <- list()
for (i in 1:length(glmQL_decided)){
  DGEs[[names(glmQL_res)[i]]] <- summary(glmQL_decided_numbs[[names(glmQL_res)[i]]])
}
DGEs_Treat <- list()
for (i in 1:length(Treat_decided)){
  DGEs_Treat[[names(Treat_decided)[i]]] <- summary(Treat_decided_numbs[[names(Treat_decided)[i]]])
}
DGEs
DGEs_Treat
```



