---
title: "Full RNAseq pipeline, from raw data to ORA/GSEA results"
author: "A. A. A. Lingua"
date: "2025-05-07"
output: html_document
---

```{r}
### Directories needed for this file
rootdir <- getwd()
Robjectsdir <- paste0(rootdir, "/Robjects")
Graphsdir<- paste0(rootdir, "/graphs")
if (!dir.exists(file.path(Graphsdir, "BioVar"))) {
  dir.create(file.path(Graphsdir, "BioVar"))
}
BVdir <- paste0(Graphsdir, "/BioVar")

#setting root directory
knitr::opts_knit$set(root.dir = rootdir)
# options(max.print = .Machine$integer.max, scipen = 999, stringsAsFactors = F, dplyr.summarise.inform = F) # avoid truncated output in R console and scientific notation
# Set seed to ensure reproducibility in any function sthat require randomness
set.seed(123456)
suppressMessages({
  library(edgeR)    # for DGE analysis
  library(pheatmap) # for heatmap graphs
  library(ggplot2)  # graphing module
  library(RColorBrewer) # for a colourful plot
  library(ggrepel) # for nice annotations
  })
```

```{r loading relevant values}
if (!exists("y")){
  y <- readRDS(file = paste0(Robjectsdir, "/y.rds"))
}
if (!exists("s2c")){
  s2c <- readRDS(file = paste0(Robjectsdir, "/s2c.rds"))
  design <- s2c$design
}


```

### Making MDS plots, heatmap(s), and BCV plots to ensure biological variation is fine

```{r MDS-all, echo=FALSE}
#####
#> To change:
#> - modify numbers of reps for, and amounts of, colours and shapes to align with data design.
#>
#####

# attributing values used in function for ease of reading
sample_count <- dim(y[["samples"]])[1]
sample1_name <- y[["samples"]][["samples"]][1]

### MDS for all samples
# producing MDSplot object, that calculates variances between samples' transcript expression data
mds <- plotMDS(y, top = 500, labels = rownames(y$samples), dim.plot = c(1,2), var.explained = TRUE)
# saving base MDS plot that uses the name of each sample as their point on the graph.
dev.copy2pdf(file = paste0(BVdir, "/MDS_all_named.pdf"))
  # top 500 'genes' to use for calculating pairwise distances.
  # dim.plot 1,2 to graph over PC1 and PC2. var.explained true adds percentage of variation
  # each principle component accounts for on their respective axis title.

# making axes labels from data in MDSplot object. 
# setting up plot customizations depending on sample list size
# grey-vehicle,red-PAO1,blue-pqsR 
cols = c(rep("grey",5), rep("red",10),rep("blue",10)) 
# square-vehicle, filled lozenge-highdose, empty lozenge-lowdose
shapes <- c(rep(0,5), rep(c(17,2),each=5,times=2)) 
group_legend <- c(c(pch=0,col="grey"), c(rep(c(pch=17,col="red"),2)),c(rep(c(pch=2,col="blue"),2)))
title <- "MDS plot of all samples"
fname <- paste0(BVdir, "/MDS_all.pdf")
# setting up plot parameters and options
# writing variance percentage of respective pc through rounding value to 2 digits and converting to percentage notation
xlab = paste0(mds[["axislabel"]]," 1 (", scales::label_percent()(round(mds[["var.explained"]][1], digits= 2)),")")
ylab = paste0(mds[["axislabel"]]," 2 (", scales::label_percent()(round(mds[["var.explained"]][2], digits= 2)),")")
# Options below initializes plot parameters. It adds margin to the right of the plot to allow for legends to be written outside of the graph
par(xpd=T, mar=par()$mar+c(0,0,0,6))
# using plot function to allow for more customization. plot initial plot with axes labels
plot(mds, xlab = xlab, ylab = ylab)
# setting colour for plot background
rect(par("usr")[1], par("usr")[3], par("usr")[2], par("usr")[4], col = "honeydew")
# writing points onto plot, with relevant colours and shapes
points(mds, col = cols, pch = shapes,bg="black")
print(shapes)
# add legends and title. Inset specifies legend location
legend('topright',
       title = c(as.expression(bquote(bold("Treatment and dose")))),
       inset=c(-0.3,0), 
       legend=c("Vehicle","PAO1 high-dose","PAO1 low-dose","PAO1ΔpqsR high-dose","PAO1ΔpqsR low-dose"), 
       pch=c(0,24,2,24,2),
       col = c("grey","red","red","blue","blue"),
       pt.bg = c("grey","red","grey","blue","grey"),
       cex = 0.7, bg = "honeydew")
title(title)
dev.copy2pdf(file = fname)

```

```{r BCV variation check}
y <- estimateDisp(y, design=design, robust=TRUE)
plotBCV(y, main = "BCV plot - All samples")
dev.copy2pdf(file = paste0(BVdir, "/BCV_plot_all.pdf"))
```

```{r heatmaps}

# Visualization of correlation of sample groups through their transcript expression values. To accommodate for large variances in transcript expression even with normalization, the values correlated are log2 of their counts per million, calculated previously as logCPM dataframe is used in other sections as well.
grouping <- s2c[,2:4]
rownames(grouping)<-s2c$sample
# Correlation of logCPM values by pearson method
allCors <- cor(logCPM, method="pearson", use="pairwise.complete.obs")
pheatmap(allCors, 
         annotation_col = grouping, 
         show_rownames = FALSE, 
         clustering_distance_rows = "correlation",
         clustering_distance_cols = "correlation",
         filename = paste0(BVdir, "/Heatmap_all_samples.pdf"))
```

```{r fitting model}
# Fitting a negative binomial generalized log-linear model to the read counts for each transcript.
fit <- glmQLFit(y, s2c$design, robust=TRUE)
plotQLDisp(fit)
saveRDS(fit, file = paste0(Robjectsdir, "/fit.rds"))
```
